### PR TITLE
Use modules properly and allow composability

### PR DESCRIPTION
**Pull request by jdesboeufs**
Hi,

I needed to detect encoding in a small frontend tool.
This library is awesome, but too huge for my use case.
I started to investigate how to cut it down.

Sometimes, it is not necessary to be able to detect all encodings available in `jschardet`.
In my case, we develop a tool for French people, so we only need to support Windows-1252/Latin-1 and UTF-8.

This PR is a code reorganization, using modules the right way (with their own dependency tree).
It's a first step before going further!

For now, it breaks compatibility for those who use actual `_debug` and `Constants.MINIMUM_THRESHOLD`. I updated the README.

This reorganization make possible writing custom detectors.
It's at the moment very quick and dirty, but if we work together I'm sure we can do a better `jschardet` `2.0` :)

See example custom code: 
https://github.com/jdesboeufs/jschardet/compare/use-modules...jdesboeufs:fr-cut

| Bundle | Size | Size (min) | Size (min+gz) |
| ---- | ---- | ---- | ---- |
| Full | 455 KB | 328 KB | 117 KB |
| Fr | 37 KB | 11 KB | 2.9 KB 🎉 💃  |

See also my independant PR #42 then #43.

### CODE REVIEW

**Comment by aadsm** [comment]
This is awesome and super needed!!! Thank you so much for doing it, I always wanted to organize it this way but could never find the time.
I’ll add a few comments <3.

File: src/escsm.js [hunk 1]
```diff
@@ -27,11 +27,9 @@
  * 02110-1301  USA
  */
 
-!function(jschardet) {
+var consts = require('./constants');
 
-var consts = jschardet.Constants;
-
-jschardet.HZ_cls = [
+exports.HZ_cls = [
```
Comment (aadsm) [inline src/escsm.js]: ´HZ_cls` and a few more are never used outside of this scope so we don’t really need to add it to `exports`. Not sure why I added it to `jschardet` at the time, but we can avoid it now.

File: src/big5prober.js [hunk 2]
```diff
@@ -27,16 +27,19 @@
  * 02110-1301  USA
  */
 
-!function(jschardet) {
+var CodingStateMachine = require('./codingstatemachine')
```
Comment (aadsm) [inline src/big5prober.js]: We can also use this reorganization to make the filenames camel case, so they match their symbols: `CodingStateMachine.js`.

File: README.md [hunk 3]
```diff
@@ -39,13 +39,13 @@ Options
 ```javascript
 // See all information related to the confidence levels of each encoding.
 // This is useful to see why you're not getting the expected encoding.
-jschardet.Constants._debug = true;
+jschardet.setLogger();
```
Comment (aadsm) [inline README.md]: This is fine, we just need to bump it to `2.0.0`.

File: src/index.js [hunk 4]
```diff
@@ -1 +1,46 @@
...
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301  USA
+ */
+
+var UniversalDetector = require('./universaldetector');
+var setLogger = require('./logger').setLogger;
+
+exports.VERSION = "1.4.1";
```
Comment (aadsm) [inline src/index.js]: We can get rid of this.

File: README.md [hunk 5]
```diff
@@ -39,13 +39,13 @@ Options
...
 
 // Default minimum accepted confidence level is 0.20 but sometimes this is not
 // enough, specially when dealing with files mostly with numbers.
 // To change this to 0 to always get something or any other value that can
 // work for you.
-jschardet.Constants.MINIMUM_THRESHOLD = 0;
+jschardet.detect(str, { minimumThreshold: 0 });
```
Comment (aadsm) [inline README.md]: 👍

File: src/index.js [hunk 6]
```diff
@@ -1 +1,46 @@
...
+    u.reset();
+    if( typeof Buffer == 'function' && buffer instanceof Buffer ) {
+        u.feed(buffer.toString('binary'));
+    } else {
+        u.feed(buffer);
+    }
+    u.close();
+    return u.result;
+}
+exports.UniversalDetector = UniversalDetector;
```
Comment (aadsm) [inline src/index.js]: We don’t need to export this I believe.

File: src/logger.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
+// By default, do nothing
+exports.log = function () {};
+
+exports.setLogger = function setLogger(fn) {
+  exports.enabled = true;
```
Comment (aadsm) [inline src/logger.js]: Instead of exporting this and add the check everywhere, we can just make the check on the log function itself (at line 10).

**Comment by danielgindi** [comment]
@aadsm is there an ETA on this? as this is very important to allow static compilation, like with `rollup`. The circular dependency with `init.js` currently is a major problem.

**Comment by aadsm** [comment]
@jdesboeufs I had forgotten about this, I'll try to take a look over the weekend.

**Comment by aadsm** [comment]
I made some changes and I just published a new version with this.

**Pull request closed without merging**